beacon-chain/state/testing/getters.go

```diff
@@ -0,0 +1,192 @@
+package testing
+
+import (
+	"sync"
+	"testing"
+
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	fieldparams "github.com/prysmaticlabs/prysm/config/fieldparams"
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/testing/assert"
+	"github.com/prysmaticlabs/prysm/testing/require"
+)
+
+func VerifyBeaconState_SlotDataRace(t *testing.T, factory getState) {
+	headState, err := factory()
+	require.NoError(t, err)
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		require.NoError(t, headState.SetSlot(0))
+		wg.Done()
+	}()
+	go func() {
+		headState.Slot()
+		wg.Done()
+	}()
+
+	wg.Wait()
+}
+
+type getStateWithCurrentJustifiedCheckpoint func(*ethpb.Checkpoint) (state.BeaconState, error)
+type clearInternalState func(state.BeaconState)
+
+func VerifyBeaconState_MatchCurrentJustifiedCheckpt(t *testing.T, factory getStateWithCurrentJustifiedCheckpoint, clear clearInternalState) {
+	c1 := &ethpb.Checkpoint{Epoch: 1}
+	c2 := &ethpb.Checkpoint{Epoch: 2}
+	beaconState, err := factory(c1)
+	require.NoError(t, err)
+	require.Equal(t, true, beaconState.MatchCurrentJustifiedCheckpoint(c1))
+	require.Equal(t, false, beaconState.MatchCurrentJustifiedCheckpoint(c2))
+	require.Equal(t, false, beaconState.MatchPreviousJustifiedCheckpoint(c1))
+	require.Equal(t, false, beaconState.MatchPreviousJustifiedCheckpoint(c2))
+	clear(beaconState)
+	require.Equal(t, false, beaconState.MatchCurrentJustifiedCheckpoint(c1))
+}
+
+func VerifyBeaconState_MatchCurrentJustifiedCheckptNative(t *testing.T, factory getStateWithCurrentJustifiedCheckpoint) {
```

Comment:
You do not need to separate out for our native state. They all follow the same interface, you can use the above test for the native state. This applies to all the below following tests.

Reply:
@nidas The Native one does not have `clear` method, that is why I indroduced this

Reply:
You can just have the native state clear method to do nothing so as to reduce test duplication here.

Reply:
If I am not understanding something from your comment, let's fix it in another PR, so this does not get in more conflicts

Reply:
Maybe it will be clearer with an example:
```go
func TestBeaconState_MatchCurrentJustifiedCheckpt(t *testing.T) {
	testtmpl.VerifyBeaconState_MatchCurrentJustifiedCheckpt(
		t,
		func(cp *ethpb.Checkpoint) (state.BeaconState, error) {
			return InitializeFromProto(&ethpb.BeaconState{CurrentJustifiedCheckpoint: cp})
		},
func(i state.BeaconState) { 
}
	)
}
```
Basically the clear function is no-op .